src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -35,62 +38,51 @@ public RouteValueDictionary()
         /// </remarks>
         public RouteValueDictionary(object values)
         {
-            var otherDictionary = values as RouteValueDictionary;
-            if (otherDictionary != null)
+            RouteValueDictionary dictionary;
+            IEnumerable<KeyValuePair<string, object>> keyValueEnumerable;
+            if ((dictionary = values as RouteValueDictionary) != null)
             {
-                if (otherDictionary.InnerDictionary != null)
-                {
-                    InnerDictionary = new Dictionary<string, object>(
-                        otherDictionary.InnerDictionary.Count,
-                        StringComparer.OrdinalIgnoreCase);
-
-                    foreach (var kvp in otherDictionary.InnerDictionary)
-                    {
-                        InnerDictionary[kvp.Key] = kvp.Value;
-                    }
-
-                    return;
-                }
-                else if (otherDictionary.Properties != null)
+                ListStorage listStorage;
+                if ((listStorage = dictionary._storage as ListStorage) != null)
                 {
-                    Properties = otherDictionary.Properties;
-                    Value = otherDictionary.Value;
+                    _storage = new ListStorage(listStorage);
                     return;
                 }
-                else
+
+                PropertyStorage propertyStorage;
+                if ((propertyStorage = dictionary._storage as PropertyStorage) != null)
                 {
+                    _storage = new PropertyStorage(propertyStorage);
                     return;
                 }
-            }
 
-            var keyValuePairCollection = values as IEnumerable<KeyValuePair<string, object>>;
-            if (keyValuePairCollection != null)
+                // If we get here, it's an EmptyStorage
+                _storage = EmptyStorage.Instance;
+                return;
+            }
+            else if ((keyValueEnumerable = values as IEnumerable<KeyValuePair<string, object>>) != null)
```

Comment:
if you have to do return; at the end of each block I think it's clearer if you just write the code as a sequence of ifs, rather than doing if else all around, i think its clearer that way.